soroban-env-host/src/events/internal.rs

```diff
@@ -0,0 +1,143 @@
+            None => None,
+        };
+        Ok(xdr::ContractEvent {
+            ext: xdr::ExtensionPoint::V0,
+            contract_id,
+            type_: self.type_,
+            body: xdr::ContractEventBody::V0(xdr::ContractEventV0 { topics, data }),
+        })
+    }
+}
+
+/// The internal representation of an `Event` that is stored in the events buffer
+/// and designed to be cheap to cloned.
+#[derive(Clone, Debug, Default)]
+pub(crate) enum InternalEvent {
+    Contract(InternalContractEvent),
+    Debug(DebugEvent),
+    #[default]
+    None,
+}
+
+impl MeteredClone for InternalEvent {}
+
+/// The events buffer. Stores `InternalEvent`s in the chronological order.
+#[derive(Clone, Default)]
+pub(crate) struct InternalEventsBuffer {
+    pub(crate) vec: MeteredVector<InternalEvent>,
+}
+
+impl InternalEventsBuffer {
+    // Records an InternalEvent
+    // Metering covered by the `MeteredVec`.
+    pub fn record(&mut self, e: InternalEvent, budget: &Budget) -> Result<(), HostError> {
+        self.vec = self.vec.push_back(e, budget)?;
+        Ok(())
+    }
+
+    fn defunct_contract_event(c: &mut InternalContractEvent) -> InternalEvent {
+        let ty: RawVal = <i32>::from(c.type_).into();
+        let id: RawVal = c.contract_id.map_or(RawVal::from_void(), |obj| obj.into());
+        let dbg = DebugEvent::new()
+            .msg("rolled-back contract event: type {}, id {}, topics {}, data {}")
+            .arg(ty)
+            .arg(id)
+            .arg(RawVal::from(c.topics))
+            .arg(c.data);
+        InternalEvent::Debug(dbg)
+    }
+
+    /// Rolls back the event buffer starting at `events`. Any `ContractEvent` will be converted
+    /// to a `DebugEvent` indicating the event has been rolled back. An additional `DebugEvent`
+    /// will be pushed at the end indicating the rollback happened.
+    // Metering covered by the `MeteredVec`
+    pub fn rollback(&mut self, events: usize, host: &Host) -> Result<(), HostError> {
+        let mut rollback_count = 0u32;
+        self.vec = self.vec.retain_mut(
+            |i, e| {
+                if i < events {
+                    Ok(true)
+                } else {
+                    if let InternalEvent::Contract(c) = e {
+                        *e = Self::defunct_contract_event(c);
+                        rollback_count += 1;
+                        Ok(true)
+                    } else {
+                        Ok(true)
+                    }
+                }
+            },
+            host.as_budget(),
+        )?;
+        self.vec = self.vec.push_back(
+            InternalEvent::Debug(
+                DebugEvent::new()
+                    .msg("{} contract events rolled back. Rollback start pos = {}")
+                    .arg(RawVal::from(rollback_count))
+                    .arg(host.usize_to_rawval_u32(events)?),
+            ),
+            host.as_budget(),
+        )?;
+        Ok(())
+    }
+
```

Comment:
@graydon Here is the events rollback logic we discussed, where a contract event will be rolled back to a debug event showing its "rolled back" status. An additional event is added in the end to indicate the roll back happened. 
I'm also thinking to maybe only include the last event if there were any contract events rolled back (i.e. `rollback_count` > 0). Otherwise we will just be seeing a not-so-meaningful message of "0 contract events rolled back..". What do you think?
cc @sisuresh

Reply:
Yeah I think it makes sense to only show this message if an event was rolled back.

Reply:
Yeah, probably only emit it if there were actual rolled-back events.

Reply:
Fixed.